File: 802.cpp
struct Solution {
    bool dfs(int x, vector<vector<int>>& graph, vector<char> &tag) {
        if (tag[x] == 1){
            return false;
        }
        if (tag[x] == 2){
            return true;
        }
        tag[x] = 1;
        for (int y : graph[x]){
            if (!dfs(y, graph, tag)){
                return false;
            }
        }
        tag[x] = 2;
        return true;
    }
    inline vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<char> tag(n, 0);
        vector<int> res;
        for (int i = 0; i < n; ++i){
            if (dfs(i, graph, tag)){
                res.push_back(i);
            }
        }
        return res;
    }
};

File: sol/802.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n = len(graph)
        tag = [0] * n
        def dfs(x):
            if tag[x] == 1:
                return False
            if tag[x] == 2:
                return True
            tag[x] = 1
            for y in graph[x]:
                if not dfs(y):
                    return False
            tag[x] = 2
            return True
        return [i for i in range(n) if dfs(i)]
